From 5a0a4d95ab15f7654babb5eebd342d08ba1ddbeb Mon Sep 17 00:00:00 2001
From: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Date: Tue, 5 Mar 2019 17:54:10 -0800
Subject: [PATCH 0538/1852] dma: xilinx: Support early callback modes

commit 0b5c9b34fa3e78ec57b3fa60c3ca97305ac5649c from
https://github.com/Xilinx/linux-xlnx.git

In the current implementation, DMA client driver can only enable or
disable the early callback. This patch enables support for early callback
modes. This means that DMA client drivers can pass various early callback
modes based on end user requirement.

The patch is using BIT(1) for existing early callback mode, and it's
backward compatible with previous definition. So the caller of this
function doesn't break with this commit.

Signed-off-by: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/dma/xilinx/xilinx_frmbuf.c | 14 +++++++-------
 include/linux/dma/xilinx_frmbuf.h  | 12 ++++++------
 2 files changed, 13 insertions(+), 13 deletions(-)

diff --git a/drivers/dma/xilinx/xilinx_frmbuf.c b/drivers/dma/xilinx/xilinx_frmbuf.c
index 6bcd686bcd6c..b80d2f6dd063 100644
--- a/drivers/dma/xilinx/xilinx_frmbuf.c
+++ b/drivers/dma/xilinx/xilinx_frmbuf.c
@@ -145,7 +145,7 @@ struct xilinx_frmbuf_tx_descriptor {
 	struct xilinx_frmbuf_desc_hw hw;
 	struct list_head node;
 	u32 fid;
-	bool earlycb;
+	u32 earlycb;
 };
 
 /**
@@ -802,7 +802,7 @@ EXPORT_SYMBOL(xilinx_xdma_set_fid);
 
 int xilinx_xdma_get_earlycb(struct dma_chan *chan,
 			    struct dma_async_tx_descriptor *async_tx,
-			    bool *enable)
+			    u32 *earlycb)
 {
 	struct xilinx_frmbuf_device *xdev;
 	struct xilinx_frmbuf_tx_descriptor *desc;
@@ -811,21 +811,21 @@ int xilinx_xdma_get_earlycb(struct dma_chan *chan,
 	if (IS_ERR(xdev))
 		return PTR_ERR(xdev);
 
-	if (!async_tx || !enable)
+	if (!async_tx || !earlycb)
 		return -EINVAL;
 
 	desc = to_dma_tx_descriptor(async_tx);
 	if (!desc)
 		return -EINVAL;
 
-	*enable = desc->earlycb;
+	*earlycb = desc->earlycb;
 	return 0;
 }
 EXPORT_SYMBOL(xilinx_xdma_get_earlycb);
 
 int xilinx_xdma_set_earlycb(struct dma_chan *chan,
 			    struct dma_async_tx_descriptor *async_tx,
-			    bool enable)
+			    u32 earlycb)
 {
 	struct xilinx_frmbuf_device *xdev;
 	struct xilinx_frmbuf_tx_descriptor *desc;
@@ -841,7 +841,7 @@ int xilinx_xdma_set_earlycb(struct dma_chan *chan,
 	if (!desc)
 		return -EINVAL;
 
-	desc->earlycb = enable;
+	desc->earlycb = earlycb;
 	return 0;
 }
 EXPORT_SYMBOL(xilinx_xdma_set_earlycb);
@@ -1164,7 +1164,7 @@ static irqreturn_t xilinx_frmbuf_irq_handler(int irq, void *data)
 
 	/* Check if callback function needs to be called early */
 	desc = chan->staged_desc;
-	if (desc && desc->earlycb) {
+	if (desc && desc->earlycb == EARLY_CALLBACK) {
 		callback = desc->async_tx.callback;
 		callback_param = desc->async_tx.callback_param;
 		if (callback) {
diff --git a/include/linux/dma/xilinx_frmbuf.h b/include/linux/dma/xilinx_frmbuf.h
index 67848abbcf15..5307113efd5b 100644
--- a/include/linux/dma/xilinx_frmbuf.h
+++ b/include/linux/dma/xilinx_frmbuf.h
@@ -133,25 +133,25 @@ int xilinx_xdma_set_fid(struct dma_chan *chan,
  *
  * @chan: dma channel instance
  * @async_tx: descriptor whose parent structure contains fid.
- * @enable: Output param - Early callback enabled
+ * @earlycb: Output param - Early callback mode
  *
  * Return: 0 on success, -EINVAL in case of invalid chan
  */
 int xilinx_xdma_get_earlycb(struct dma_chan *chan,
 			    struct dma_async_tx_descriptor *async_tx,
-			    bool *enable);
+			    u32 *earlycb);
 
 /**
  * xilinx_xdma_set_earlycb - Enable/Disable early callback
  * @chan: dma channel instance
  * @async_tx: dma async tx descriptor for the buffer
- * @enable: Flag to enable or disable early callback for descriptor.
+ * @earlycb: Enable early callback mode for descriptor
  *
  * Return: 0 on success, -EINVAL in case of invalid chan
  */
 int xilinx_xdma_set_earlycb(struct dma_chan *chan,
 			    struct dma_async_tx_descriptor *async_tx,
-			    bool enable);
+			    u32 earlycb);
 #else
 static inline void xilinx_xdma_set_mode(struct dma_chan *chan,
 					enum operation_mode mode)
@@ -192,14 +192,14 @@ static inline int xilinx_xdma_set_fid(struct dma_chan *chan,
 
 static inline int xilinx_xdma_get_earlycb(struct dma_chan *chan,
 					  struct dma_async_tx_descriptor *atx,
-					  bool *enable)
+					  u32 *earlycb)
 {
 	return -ENODEV;
 }
 
 static inline int xilinx_xdma_set_earlycb(struct dma_chan *chan,
 					  struct dma_async_tx_descriptor *atx,
-					  bool enable)
+					  u32 earlycb)
 {
 	return -ENODEV;
 }
-- 
2.31.1

